### ISSUE TITLE
Panic when receiving too many fds?

### ISSUE THREAD

**Issue by chadaustin** [standard-ai/sendfd#14]
Regarding the comment and `unreachable!()` in https://github.com/standard-ai/sendfd/blob/69994c1b6d0ceb72e39afd4854694b38d6132861/src/lib.rs#L155-L164

I am using `sendfd` in a new project. While researching the actual semantics of fd-passing, I found Kenton Varda's excellent summary of the subtleties and dangers: https://gist.github.com/kentonv/bc7592af98c68ba2738f4436920868dc

In particular:

> The CMSG_SPACE() macro is intended to help you decide how much space to allocate to receive an ancillary message. It rounds up its calculation to the next word boundary. Unfortunately, on 64-bit systems, this means you will always end up with enough space for an even number of file descriptors. If you were expecting just one FD, you'll end up with enough buffer space to receive two. You MUST check whether you received two and close the second one, otherwise, again, an attacker can fill up your FD table.

That implies that calling recvmsg expecting one fd will panic if two are actually sent.

I have not tried to reproduce this issue in a test.

**Comment by nagisa** [comment]
I would accept a PR fixing this. It has been a while since I worked on this crate, so I may misremember some things, but IIRC one of the reasons we were careful around this area of the code was specifically because we didn't really want to make a decision that dropping unexpected file descriptors in the library is the right thing to do. In fact, if we cannot really receive exactly up-to the number of file descriptors as allowed by the file descriptor buffer, I would consider the API for this library to be wrong.

I wonder if it would be possible to set the `msg_controllen` to be exactly right despite what the `CMSG_SPACE` returns :thinking:

**Comment by roblabla** [comment]
I am seeing a similar problem on macos. In my case, when the sender sends more FD than the receiver can receive, it causes a crash in the receiver. Here's a simple reproducer:

```rust
use std::fs::File;
use std::os::fd::AsRawFd;
use std::os::unix::net::UnixStream;

use sendfd::{RecvWithFd, SendWithFd};

fn main() {
    let (tx, rx) = UnixStream::pair().unwrap();

    // A single sendmsg carrying four descriptors.
    let files: Vec<File> = (0..4).map(|_| File::open("/dev/null").unwrap()).collect();
    let raw: Vec<i32> = files.iter().map(|f| f.as_raw_fd()).collect();
    tx.send_with_fd(b"x", &raw).unwrap();

    // Receive with room for one. sendfd sizes its cmsg buffer from this slice.
    let mut bytes = [0u8; 8];
    let mut fds = [0i32; 1];
    let (n, fd_count) = rx.recv_with_fd(&mut bytes, &mut fds).unwrap();

    println!("returned {} bytes, {} fds", n, fd_count);
}
```

Here's the output:

```
thread 'main' panicked at sendfd-0.4.4/src/lib.rs:164:25:
internal error: entered unreachable code
```

---

This is really problematic, as it makes it very simple to DOS a service that exposes an API using sendfd that expects fd passing on macos. I think it happens because `rawfd_count` is derived from the kernel's `cmsg_len`, and XNU reports the sender's full descriptor count on a truncated receive (setting `MSG_CTRUNC`) rather than the number that fit. This is _probably_ related to the fact that, on XNU (and older versions of FreeBSD), the excess descriptors that don't fit in the recv buffer are nevertheless added to the process's file descriptor table, leaking them.

**Comment by nagisa** [comment]
> I think it happens because rawfd_count is derived from the kernel's cmsg_len, and XNU reports the sender's full descriptor count on a truncated receive (setting MSG_CTRUNC) rather than the number that fit. This is probably related to the fact that, on XNU (and older versions of FreeBSD), the excess descriptors that don't fit in the recv buffer are nevertheless added to the process's file descriptor table, leaking them.

Is that really what's happening? If so, then the behaviour is no less problematic than the panic, and sounds impossible to deal with? If true, it would still allow to easily DOS the receiver by exhausting its fd space with no way for the receiver to deal with that.

The `CMSG_SPACE` problem identified in OP, on the other hand, could be a relatively easy fix. I posted a PR at #23 for this. Mind testing it out?

---

FWIW I personally only ever used FD passing with the datagram variant of the socket. With Stream the data and ancillary data are their own independent streams with no synchronization guarantees in either direction. Which then leads directly to the problem where the application might expect to deal in fixed size messages with a single fd attached to them, but actually receives multiple backlogged file descriptors.

**Comment by nagisa** [comment]
Yeah, looks like macOS is still doing the buggy thing. I have no idea how to address this on the library side other than to try use `mach_msg` as mentioned here: https://gist.github.com/kentonv/bc7592af98c68ba2738f4436920868dc?permalink_comment_id=5240861#gistcomment-5240861 or to implement library-side buffering of the received fds. I don't have time for either option right now.

As usual PRs fixing this are welcome.
